internal/xds/rbac/rbac_engine_test.go

```diff
@@ -0,0 +1,625 @@
+		},
+	}
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			_, err := NewEngine(test.rbacConfig)
+			if test.successfullyConstruct {
+				// If the config is supposed to successfully construct, then
+				// there should be no error raised.
+				if err != nil {
+					t.Fatalf("RBAC Engine should have successfully constructed and not returned an error")
+				}
+			} else {
+				// If the config has a logical problem, then there should be an
+				// error returned.
+				if err == nil {
+					t.Fatalf("RBAC Engine should not have constructed and there should have been an error returned")
+				}
+			}
+		})
+	}
+}
+
+// TestRBACEngine tests the RBAC Engine by configuring the engine in different
+// scenarios. After configuring the engine in a certain way, this test pings the
+// engine with different kinds of data representing incoming RPC's, and verifies
+// that it works as expected.
+func (s) TestRBACEngine(t *testing.T) {
+	tests := []struct {
+		name        string
+		rbacConfig  *v3rbacpb.RBAC
+		rbacQueries []struct {
+			evaluateArgs           *RPCData
+			wantMatchingPolicyName string
+			matchingPolicyFound    bool
+		}
+	}{
+		// TestSuccessCaseAnyMatch tests an RBAC Engine instantiated with a
+		// config with a policy with any rules for both permissions and
+		// principals, meaning that any data about incoming RPC's that the RBAC
+		// Engine is queried with should match that policy.
+		{
+			name: "TestSuccessCaseAnyMatch",
+			rbacConfig: &v3rbacpb.RBAC{
+				Policies: map[string]*v3rbacpb.Policy{
+					"anyone": {
+						Permissions: []*v3rbacpb.Permission{
+							{Rule: &v3rbacpb.Permission_Any{Any: true}},
+						},
+						Principals: []*v3rbacpb.Principal{
+							{Identifier: &v3rbacpb.Principal_Any{Any: true}},
+						},
+					},
+				},
+			},
+			rbacQueries: // Any incoming RPC should match with the anyone policy
+			[]struct {
+				evaluateArgs           *RPCData
+				wantMatchingPolicyName string
+				matchingPolicyFound    bool
+			}{
+				{evaluateArgs: &RPCData{
+					FullMethod: "some method",
+				},
+					wantMatchingPolicyName: "anyone",
+					matchingPolicyFound:    true,
+				},
+				{evaluateArgs: &RPCData{
+					DestinationPort: 100,
+				},
+					wantMatchingPolicyName: "anyone",
+					matchingPolicyFound:    true,
+				},
+			},
+		},
+		// TestSuccessCaseSimplePolicy is a test that tests a simple policy that
+		// only allows an rpc to proceed if the rpc is calling a certain path
+		// and port.
+		{
+			name: "TestSuccessCaseSimplePolicy",
+			rbacConfig: &v3rbacpb.RBAC{
+				Policies: map[string]*v3rbacpb.Policy{
+					"localhost-fan": {
+						Permissions: []*v3rbacpb.Permission{
+							{Rule: &v3rbacpb.Permission_DestinationPort{DestinationPort: 8080}},
+							{Rule: &v3rbacpb.Permission_UrlPath{UrlPath: &v3matcherpb.PathMatcher{Rule: &v3matcherpb.PathMatcher_Path{Path: &v3matcherpb.StringMatcher{MatchPattern: &v3matcherpb.StringMatcher_Exact{Exact: "localhost-fan-page"}}}}}},
```

Comment:
Optional: You could split this into multiple lines if that makes it more readable.

Reply:
I used to have it like that split. I like your heuristic of a single field, having it {singleField: x} and wrapped in one line.